### DISCUSSION TITLE
Location of important metadata files (such as the Dir ID)

### DISCUSSION THREAD

**Discussion by overheadhunter** [encryption-alliance/unified-vault-format#9] [Ideas]
I already wrote it in a comment on a different discussion, but I think it is important enough to start a separate topic:

Suggestion: Store files such as the one for the directory ID or the general config files in a way that they can be loaded from remote storage in O(1).

Captain obvious might ask why I think this is worth noting. Here is why: As soon as the file storage (be it cloud or local) has some pagination mechanism, finding the relevant file is a O(n) task. This is especially relevant, since we aim for a solution that has both mobile devices as well as syncing software in mind that may connect to slow remotes.

Currently I see two approaches:

## 1. Separate Metadata from Payload
```
dir/
├─ subdir/
│  ├─ containing
│  ├─ all
│  ├─ other
│  ├─ nodes
├─ dirid.file
```

## 2. Rely on Ordering
```
dir/
├─ _dirid.file
├─ all
├─ nodes
├─ other
```

## Comparison

| | 1. Separate Metadata from Payload | 2. Rely on Ordering |
|---|---|---|
| Simpler dir structure (fulfils requirement of a 1:1 mapping between ciphertext:cleartext) | ✖️  | ✔️ |
| Restoring broken dirs independent of other dirs | ✖️ | ✔️ |
| Independent of locale settings | ✔️ | ✖️ |
| Works if dir listing doesn't obey any order | ✔️ | ✖️ |
| True O(1) even in unpaginated long-running dir listings | ✔️ | ✖️ |

## Suggestion

Mix both: Use Method 1 for vault config, Method 2 for dir-specific metadata.

```
vault/
├─ root/
│  ├─ _dirid.file
│  ├─ all
│  ├─ nodes/
│  │  ├─ _dirid.file
│  │  ├─ foo
│  ├─ other
├─ config.file
```

This way, we can load relevant general vault config without listing/reading vault contents. But inside of encrypted dirs we have the possibly slower (assuming bad remote dir listing) but more resilient structure (availability as a security target wins over performance).

**Comment by rfjakob** [comment]
I did not know about the pagination problem. But why do you need the dir listing for opening a file with a known name?

**Reply by overheadhunter** [reply]
> But why do you need the dir listing for opening a file with a known name?

We don't necessarily. 😉 I just had a talk with @tobihagemann about this, turns out it is merely an optimization. There are "ID-based" cloud providers such as Google Drive, pCloud or OneDrive (and possibly others), that don't allow downloading file contents for a given path directly. Instead the path needs to be resolved to an ID, which is then used to make the download request.

If this ID can be obtained on the first page of a dir listing, we can safe one additional request.

**Comment by tobihagemann** [comment]
Just wanted to mention that we should avoid `_` (underscore) as a prefix, see: https://github.com/cryptomator/cloud-access-swift/issues/19
